### ISSUE TITLE
Direnv integration broken with elixir

### ISSUE THREAD

**Issue by happysalada** [hlissner/doom-emacs#4082]
**What did you expect to happen?**
In an elixir project with a valid shell.nix setup, direnv fails to be executed properly.
When saving a file, the formatter and linter should run without errors 

**What actually happened?**
running the formatter will give the following error
`** (Mix) Could not find an SCM for dependency :phoenix from Union.Mixfile`
In order to run the formatter and linter, two executables need to be there. `elixir` and `mix`
It looks like the direnv plugin is enabling the `elixir` executable to be found, otherwise that error wouldn't be there.
However somehow it can't find mix.

My shell.nix has some envirnment variables enabled for this something like
```
  export MIX_HOME=$PWD/.nix-mix
  export HEX_HOME=$PWD/.nix-mix
  export PATH=$MIX_HOME/bin:$PATH
  export PATH=$HEX_HOME/bin:$PATH
```

It might be that those envirnment variables are not seen by the direnv plugin.

I don't know how to debug this further, but with some instructions I would be happy to look more into it

**Additional details:**
- If I have the formatter running automatically, but it can be run with `SPC c f` (it's calling flycheck I believe)
- Let me know if you need any more details

**Steps to reproduce:**
1. in an elixir project with a shell.nix
2. add a hooks to your shell.nix with the following
```
''
  export MIX_HOME=$PWD/.nix-mix
  export HEX_HOME=$PWD/.nix-mix
  export PATH=$MIX_HOME/bin:$PATH
  export PATH=$HEX_HOME/bin:$PATH
''
```
3. have elixir mode enabled in your init.el for doom
4. try to run the formatter with `SPC c f` in an .ex file

**System information:**
<details>

```
SYSTEM  type       darwin
        config     x86_64-apple-darwin17.7.0
        shell      ~/.nix-profile/bin/fish
        uname      Darwin 19.6.0 Darwin Kernel Version 19.6.0: Mon Aug 31 22:12:52 PDT 2020; root:xnu-6153.141.2~1/RELEASE_X86_64 x86_64
        path       (/run/current-system/sw/bin/ /nix/var/nix/profiles/default/bin/ ~/.nix-profile/bin/ /usr/local/opt/fzf/bin/ /usr/local/bin/ /usr/bin/ /bin/ /usr/sbin/ /sbin/ /run/current-system/sw/bin/ /nix/var/nix/profiles/default/bin/ ~/.nix-profile/bin/ /usr/local/opt/fzf/bin/ /usr/local/bin/ /usr/bin/ /bin/ /usr/sbin/ /sbin/)
EMACS   dir        /nix/store/sg0h168klzgk6airmp5s2mggvmv02qfv-doom-emacs/
        version    27.1
        build      Oct 12, 2020
        buildopts  --prefix=/nix/store/k21akjr53iagwvapw8622im5bddzlh54-emacs-27.1 --disable-build-details --with-modules --with-ns --disable-ns-self-contained CFLAGS=-DMAC_OS_X_VERSION_MAX_ALLOWED=101200
        features   RSVG GLIB NOTIFY KQUEUE ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS MODULES THREADS JSON PDUMPER GMP
        traits     (server-running)
DOOM    dir        /nix/store/s1q8zih1646h2xqrv2nw4q671vkwzd2c-doom-private/
        version    2.0.9
        build      fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
        elc-files  0
        modules    (:completion company ivy :ui doom doom-dashboard doom-quit hl-todo ligatures modeline ophints (popup +defaults) vc-gutter vi-tilde-fringe workspaces :editor (evil +everywhere) file-templates fold (format +onsave) multiple-cursors rotate-text snippets :emacs dired electric undo vc :checkers syntax spell :tools direnv (eval +overlay) lookup magit rgb :os macos :lang (clojure +lsp) (elixir +lsp) emacs-lisp json javascript markdown nix org (rust +lsp) sh web :config (default +bindings +smartparens))
        packages   ((magit-delta))
        unpin      (n/a)
        elpa       (n/a)
```
</details>

**Comment by hlissner** [comment]
> ** (Mix) Could not find an SCM for dependency :phoenix from Union.Mixfile' In order to run the formatter and linter, two executables need to be there. elixirandmixIt looks like the direnv plugin is enabling theelixir` executable to be found, otherwise that error wouldn't be there.
> However somehow it can't find mix.

Due to poor formatting it's difficult to tell where the error message ends and your comments begin.

> it can be run with SPC c f (it's calling flycheck I believe)

`SPC c f` invokes the formatter. Nothing to do with flycheck. But if the formatter works this way, I don't understand why it wouldn't work when you save.

> When saving a file, the formatter and linter should run without errors

Is the formatter or the linter giving you this error?

-----

nix-shell into your dir environment. Does `which mix` produce the correct path to the executable? How about `which elixir`? If they don't, the problem is a misconfigured shell.nix or .envrc.

Then try this in emacs (while a file inside this direnv is open): `M-x eval-expression RET (executable-find "mix")`. Does that product a (correct) path?

**Comment by happysalada** [comment]
Thank you so much for coming back on this so fast!

Sorry, about the formatting error, I fixed it in my previous comment.

The formatter is giving the error then.

which mix and which elixir do give me the correct path to the executable.
`/nix/store/j160sr53vhkkxkfk1jqjj5mfnml3jznz-elixir-1.10.4/bin/mix`

the result of the command you gave me gives the same result as if I was running it outside of that nix-shell. It gives me
`/Users/raphael/.nix-profile/bin/mix`

Perhaps then, I misconfigured it? It looks like the shell is not being loaded.
I uncommented the `direnv` in the init.el file.
I checked [the doc](https://github.com/hlissner/doom-emacs/tree/develop/modules/tools/direnv) and it didn't say anything else.

Let me know if there is anything I can do to help.

**Comment by hlissner** [comment]
After uncommenting `direnv`, did you run `doom sync` on the command line and restart Emacs?

Does your project have an `.envrc` file with `use nix` in it?

If it does, when you opened your elixir file, did you do `M-x envrc-allow` to whitelist the directory? (Or run `direnv allow .` in the directory)?

If you did, after you've opened a file in the direnv'd directory check the \*Messages\* buffer for warnings or errors (press `SPC h e` or `C-h e`)

**Comment by happysalada** [comment]
Thanks again for being so responsive.

I had already run `direnv allow .` in the directory itself. I have also verified that it is working. When I cd in the directory I can run `mix` without problem.

checking the errors in the buffers gives me the following output

```
Running direnv in /Users/raphael/Documents/Projects/gillithekid/union/...
Direnv succeeded in /Users/raphael/Documents/Projects/gillithekid/union/
(No changes need to be saved)
Formatting error
Suspicious state from syntax checker elixir-credo: Flycheck checker elixir-credo returned 1, but its output contained no errors: Could not find Hex, which is needed to build dependency :phoenix
Shall I install Hex? (if running non-interactively, use "mix local.hex --force") [Yn] ** (Mix) Could not find an SCM for dependency :phoenix from Union.Mixfile
```

It's really weird, it says that direnv succeeds, but the fact that it can't find mix tells me that it doesn't.

I thought that perhaps my system had two different `direnv` executables, so I tried running `which direnv` in my directory and `(executable-find "direnv")` in doom, but both path were the same.

**Comment by hlissner** [comment]
It runs instantly because it only runs it once, when the buffer is opened, then saves it in buffer-local state.

**Comment by happysalada** [comment]
Thanks for the information.

checking with `getenv`, `HEX_HOME` is not set.

If there is anything I can have a look to debug further, please let me know. I'm interested in helping.

**Comment by hlissner** [comment]
```
Running direnv in /Users/raphael/Documents/Projects/gillithekid/union/...
Direnv succeeded in /Users/raphael/Documents/Projects/gillithekid/union/
```

From this, it appears direnv was successfully run but the state isn't being injected into the current buffer.

Unfortunately, I cannot reproduce this with any of my direnv environments, so I can only imagine it's a local configuration issue. Can you attach/pastebin your `shell.nix` and `.envrc` files so I can try your env verbatim?

**Comment by happysalada** [comment]
this is what I have

```
with import <nixpkgs-unstable> { };

let
  # define packages to install with special handling for OSX
  basePackages =
    [ git beam.packages.erlangR23.elixir_1_10 nodejs-14_x yarn postgresql_13 ];

  inputs = basePackages ++ lib.optional stdenv.isLinux inotify-tools
    ++ lib.optionals stdenv.isDarwin
    (with darwin.apple_sdk.frameworks; [ CoreFoundation CoreServices ]);

  # define shell startup command
  hooks = import ./hooks.nix;

in mkShell {
  buildInputs = inputs;
  shellHook = hooks;
}
```

and then in hooks.nix I have
```
''
  # this allows mix to work on the local directory
  mkdir -p $PWD/.nix-mix
  mkdir -p $PWD/.nix-hex
  export MIX_HOME=$PWD/.nix-mix
  export HEX_HOME=$PWD/.nix-mix
  export PATH=$MIX_HOME/bin:$PATH
  export PATH=$HEX_HOME/bin:$PATH
  mix local.hex --if-missing
  export LANG=en_US.UTF-8
  export ERL_AFLAGS="-kernel shell_history enabled"

  # postges related
  export PGDATA="$PWD/db"
''
```

Let me know if there is anything I else I can do

**Comment by hlissner** [comment]
Where's the .envrc file?

**Comment by happysalada** [comment]
Sorry It completely slipped me. It's the tradional one.

```
use_nix
```

**Comment by hlissner** [comment]
Very strange. I set up a directory with those files, opened a file in that directory in Emacs, and `$HEX_HOME` was set in the buffer's environment.

Are you on the latest commit of Doom?

**Comment by happysalada** [comment]
I'm on this commit https://github.com/hlissner/doom-emacs/commit/1456108d5bef89195b2b63f962e0ae9d24f4c652
I did not see older commits relating to direnv, but let me update anyways. I hope it wasn't because this was an old commit.

**Comment by happysalada** [comment]
I tried using the commit https://github.com/hlissner/doom-emacs/commit/8ba2f15c71cea8b3cf20c097128144cbbbd5cf53 and the problem persist

I am using https://github.com/vlaci/nix-doom-emacs to install doom-emacs. That could be why.

If it works for you then the problem must come from there. It's not fair to keep you engaged in this, closing this for now.

Thanks for your time!

**Comment by hlissner** [comment]
While fixing #3891 I realized there was a bug with Doom not triggering certain hooks at startup. For example, if you did `emacs -nw somefile.js`, the LSP server for js2-mode would not start. `envrc-mode` (what we rely on to trigger direnv) relies on those hooks as well.

When you tested the file, were you opening the file straight from the command line? e.g. `emacs ~/project/file.ex`

Or did you go the scenic route? Start Emacs, `SPC .` (or `C-x C-f`), then select ~/project/file.ex`?

**Comment by happysalada** [comment]
If I open emacs from inside the project directory then I have the proper variables in my environment (checked using `getenv`)

if I open straight from the command line, it works as well (`emacs ./union` where union is a folder. Then I can change to a file and the environment variables are correctly set)

The way I use it though is the following
use `emacs .` inside of my home directory
then `SPC p p` then select my `union` project then select a file to load.
(I've also checked just selecting a particular file with `SPC .` and it doesn't work either)

I've checked in my $HOME directory and I don't have a conflicting `.direnv` folder.

Could it be a hook that is not being run then?
